Add web sockets support

npradeep357 opened the pull request:
This pull request will enable PYLSP to run in Web Sockets configuration and can handle multiple connections in parallel. 
The Idea was to handle ~~multiple~~ parallel editing sessions using single process. This is achieved by creating new instances of PYLSP ~~class~~ Object for each WS connection in a single server process.

Fixes #117 

As of now this PR only supports text based messaging over WS. Binary messaging support will be added in future.

This PR uses ~~autobahn websockets with twisted framework~~ [Websockets Library](https://websockets.readthedocs.io/en/stable/)
~~There is asyncio based framework but twisted is used because PYLSP is based on callbacks and asyncio is not compatible with it.~~

referred docs: 
~~https://autobahn.readthedocs.io/en/latest/websocket/programming.html
https://autobahn.readthedocs.io/en/latest/installation.html~~
https://websockets.readthedocs.io/en/stable/

related dependencies are added to ~~setup.py~~ setup.cfg
related usage documentation is added to README

Commit:
Added support for Web sockets based server over PYLSP.
Using 'Autobahn' with 'twisted' to provied Web Sockets

Commit:
Updated readme with ws configuration usage

npradeep357 committed:
Update README.md

ccordoba12 reviewed (commented):
Hey @npradeep357, thanks a lot for your contribution! My main comment for now is it's possible to add some tests for this new functionality, given that right now I see none.

Besides that, I'd like to ask @krassowski what he thinks about this one.

npradeep357 commented:
@ccordoba12, Thank you, will add tests for this.

Also, regarding the checks, there are Linux, Mac and Windows tests failure which says the version mentioned is not available for Py 3.6. I took the latest version of autobahn from PyPI which is [21.11.1](https://pypi.org/project/autobahn/).

But the log in the tests shows only until 21.2.1 which is an older version released in Feb 2021. ref: [autobahn PyPI history](https://pypi.org/project/autobahn/#history)

Suggest me if I have to move to older version or is there something I'm missing for Py 3.6?

npradeep357 commented:
After some digging, I found out that autobahn is only supported from Python 3.7+.

Is there a reason that PYLSP is supported from Python 3.6+? Can we move PYLSP to Python 3.7+? or should I downgrade the autobahn library which I used?

ccordoba12 commented:
> Is there a reason that PYLSP is supported from Python 3.6+?

There is because Python 3.6 is still a supported Python version. However, since it's going to reach end-of-life this month, we can drop support for it now.

Please do that in a different pull request though, i.e. increase the Python version in our `setup.py` and remove the Python 3.6 slots on Github actions. After that you could rebase this PR so that your tests pass.

Commit:
Updated pylint checks

npradeep357 committed:
Fixed pylint issues

npradeep357 committed:
Updated trailing spaces issues

Commit:
Fixed pycodestyle checks

Commit:
fixed issue with on close misisng parameters

npradeep357 commented:
Hey @ccordoba12, I have created a separate PR for version change as mentioned. Can you review and approve #130 when you find some time. Thank you

krassowski commented on README.md:
````diff
@@ -21,6 +21,27 @@ If the respective dependencies are found, the following optional providers will
 - [autopep8](https://github.com/hhatto/autopep8) for code formatting
 - [YAPF](https://github.com/google/yapf) for code formatting (preferred over autopep8)
 
+PYLSP can now be used in Web Sockets based configuration as follows:
+
+```
+pylsp --ws --host [host/ip] --port [port]
+```
+
+The following libraries are required for Web Sockets support:
+- [autobahn-twisted](https://github.com/crossbario/autobahn-python) for PYLSP Web sockets using autobahn in twisted configuration. refer [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details
+
+you can install these dependencies with below command:
+
+```
+pip install autobahn[twisted,accelerate]>=21.11.1
+```
+
+or with this:
+
+```
+pip install 'python-lsp-server[autobahn-twisted]'
+```
+
````
Let's move this block to the end of `Configuration` section -  an average user does not care what method is used.

krassowski commented on README.md:
```diff
@@ -21,6 +21,27 @@ If the respective dependencies are found, the following optional providers will
 - [autopep8](https://github.com/hhatto/autopep8) for code formatting
 - [YAPF](https://github.com/google/yapf) for code formatting (preferred over autopep8)
 
+PYLSP can now be used in Web Sockets based configuration as follows:
```
I think we do not use PYLSP in capitals elsewhere so maybe use the full name?

krassowski replied:
```suggestion
Python LSP Server can communicate over WebSockets when configured as follows:
```

krassowski commented on README.md:
````diff
@@ -21,6 +21,27 @@ If the respective dependencies are found, the following optional providers will
+- [autobahn-twisted](https://github.com/crossbario/autobahn-python) for PYLSP Web sockets using autobahn in twisted configuration. refer [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details
+
+you can install these dependencies with below command:
+
+```
+pip install autobahn[twisted,accelerate]>=21.11.1
+```
+
+or with this:
````
Let's keep it simple and easier to maintain.

```suggestion
```

krassowski commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +91,114 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+    # pylint: disable=import-outside-toplevel
+
+    # imports needed only for websocket based server
+    from autobahn.twisted.websocket import WebSocketServerProtocol, WebSocketServerFactory
+    from twisted.internet import reactor
```
You could wrap these with try-except and re-raise with instructions for installing websocket requirements, something like "Please run `pip install 'python-lsp-server[autobahn-twisted]` or equivalent".

krassowski commented on setup.py:
```diff
@@ -65,6 +66,7 @@ def get_version(module='pylsp'):
         'pylint': ['pylint>=2.5.0'],
         'rope': ['rope>0.10.5'],
         'yapf': ['yapf'],
+        'autobahn-twisted': ['autobahn[twisted,accelerate]>=21.11.1'],
```
I wonder if the extras should be renamed to `websockets` so it is not tied to the current implementation if we need to change it in the future.

Commit:
- moved websockets information to end of configuration
- renamed extras - autobahn-twisted to websockets
- added more info on the import error

npradeep357 commented:
@krassowski made the suggested changes.

krassowski commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +91,117 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+
+    factory = WebSocketServerFactory(f"ws://{bind_addr}:{port}")
+    factory.protocol = MyServerProtocol
+    # factory.setProtocolOptions(maxConnections=2)
```
Please remove commented code, or explain why it is commented.

krassowski commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +91,117 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+        else:
+            self._jsonrpc_stream_writer = None
+
+        # if Consumer is None, It is assumed that the default streams based approach is being used
```
```suggestion
        # if Consumer is None, it is assumed that the default streams-based approach is being used
```

krassowski commented:
Thanks! Looks fine by me, only minor comments now.

Commit:
removed unnecessary comments, updated comment

krassowski reviewed (approved):
Broadly speaking it looks ok for me. Not sure about handling of the binary messages - should it emit a warning when one is received?

npradeep357 commented:
> Broadly speaking it looks ok for me. Not sure about handling of the binary messages - should it emit a warning when one is received?

Binary messages format, in general, will not be used in LSP use case as the protocol requires to send JSON messages over sockets. I assume there won't be a client which uses Binary format.

npradeep357 commented:
@krassowski, Not sure if this the right place, This PR requires #130 to be merged before merging this. Can you please review it and approve so that it can be merged before this.

Thank you.

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

npradeep357 commented:
@ccordoba12, can this also be added to v1.4.0 milestone?

ccordoba12 commented:
I'd still like to see some tests for this. See for instance:

https://stackoverflow.com/questions/40884504/pytest-testing-a-websocket-connection

npradeep357 commented:
Hi @ccordoba12, I'm still figuring out approach to complete the test case. I don't think this will go into v1.4.0. May can we move this out of the milestone for now?

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

ccordoba12 commented:
No problem. Hope you could resume this work soon.

npradeep357 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into python-lsp-develop

npradeep357 committed:
Merge branch 'python-lsp-develop' into develop

ccordoba12 commented on setup.cfg:
```diff
@@ -31,6 +31,7 @@ all =
     pylint>=2.5.0
     rope>=0.10.5
     yapf
+    autobahn[twisted,accelerate]>=21.11.1
```
I suggest to remove this line because Twisted is a big dependency and it's not really necessary for the server to work. If people want ẁebsocket functionality, they can use the `[websocket]` extra requirement directly.

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +91,116 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+    try:
+        from autobahn.twisted.websocket import WebSocketServerProtocol, WebSocketServerFactory
+        from twisted.internet import reactor
+        import ujson as json
```
Please move the `ujson` import to the beginning of the file because it's a main dependency of this project, so you don't need to check for its presence.

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +91,116 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+        else:
+            self._jsonrpc_stream_writer = None
+
+        # if Consumer is None, it is assumed that the default streams-based approach is being used
```
```suggestion
        # if consumer is None, it is assumed that the default streams-based approach is being used
```

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

npradeep357 committed:
Updated review comments

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

ccordoba12 commented on README.md:
````diff
@@ -89,6 +89,21 @@ As an example, to change the list of errors that pycodestyle will ignore, assumi
+```
+
+The following libraries are required for Web Sockets support:
+- [autobahn-python](https://github.com/crossbario/autobahn-python) for Python LSP Server Web sockets using autobahn with twisted configuration. refer [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details
````
```suggestion
The [autobahn-python](https://github.com/crossbario/autobahn-python) library is required for Web Sockets support. Refer to the [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details.
```

ccordoba12 commented on README.md:
```diff
@@ -89,6 +89,21 @@ As an example, to change the list of errors that pycodestyle will ignore, assumi
+The following libraries are required for Web Sockets support:
+- [autobahn-python](https://github.com/crossbario/autobahn-python) for Python LSP Server Web sockets using autobahn with twisted configuration. refer [Autobahn installation](https://autobahn.readthedocs.io/en/latest/installation.html) for more details
+
+you can install these dependencies with below command:
```
```suggestion
You can install this dependency with command below:
```

ccordoba12 replied:
Please address this suggestion.

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +92,115 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+    class MyServerProtocol(WebSocketServerProtocol):
+
+        def handler(self, message):
+            """Handler to send responses of  processed requests to respective web socket clients"""
```
```suggestion
            """Handler to send responses of processed requests to the respective web socket clients."""
```

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +92,115 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
+
+        def onMessage(self, payload, isBinary):
+            if isBinary:
+                # binary message support will be provided in future
```
```suggestion
                # Binary message support will be provided in the future.
```

ccordoba12 commented on pylsp/python_lsp.py:
```diff
@@ -91,32 +92,115 @@ def start_io_lang_server(rfile, wfile, check_parent_process, handler_class):
         self._jsonrpc_stream_reader.listen(self._endpoint.consume)
 
+    def consume(self, message):
+        """Entry point for consumer based server. Alternate to stream listeners"""
```
```suggestion
        """Entry point for consumer based server. Alternative to stream listeners."""
```

ccordoba12 replied:
This suggestion is also still unaddressed. Please apply it.

ccordoba12 reviewed (commented):
@npradeep357, thanks for your patience with this! Last review and then this should be ready (we can take a look at adding tests later).

npradeep357 commented:
> @npradeep357, thanks for your patience with this! Last review and then this should be ready (we can take a look at adding tests later).

@ccordoba12 I'm planning to change the websockets library to reduce 3rd party library footprint. If it can wait a week or two, I will update the PR.

ccordoba12 commented:
Sure, no problem with that.

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

npradeep357 committed:
Websockets (#4)

* Added websockets based implementation to reduce libraries and 3rd party dependencies

* Removed message encode and decode as they are no longer needed

Co-authored-by: Pradeep Neerukonda <pneeruko@opentext.com>

npradeep357 committed:
Update README.md

npradeep357 committed:
removed triling whitespaces

npradeep357 committed:
removed space between parameters

npradeep357 commented:
@ccordoba12, @krassowski, removed autobahn and twisted libraries and using websockets library to reduce libraries footprint.
please review, approve and merge this.

I have tested it my application and everything's working as expected..

websockets: https://websockets.readthedocs.io/en/stable/

npradeep357 committed:
Updated debug log

npradeep357 committed:
changed logic to use single threadpool excutor

npradeep357 commented:
@ccordoba12 @krassowski could you please review the latest changes.

npradeep357 committed:
Merge branch 'develop' of https://github.com/python-lsp/python-lsp-server into python-lsp-develop

# Conflicts:
#	setup.cfg

npradeep357 committed:
Merge branch 'python-lsp-develop' into develop

npradeep357 committed:
updated websockets version to 10.3

npradeep357 commented:
@ccordoba12, can we merge this?

ccordoba12 reviewed (commented):
@npradeep357, thanks for your patience with this! There are still two unaddressed suggestions from my previous review. After you apply them, I'll merge your PR.

npradeep357 committed:
Merge branch 'python-lsp:develop' into develop

npradeep357 committed:
Addressing comments

npradeep357 commented:
Hi @ccordoba12, Somehow these suggestions got overlooked. I have addressed them now. Thank you

ccordoba12 reviewed (approved):
Thanks for your contribution @npradeep357!
